samza-sql/src/main/java/org/apache/samza/sql/translator/JoinTranslator.java

```diff
@@ -0,0 +1,279 @@
+import org.apache.calcite.rel.core.JoinRelType;
+import org.apache.calcite.rel.logical.LogicalJoin;
+import org.apache.calcite.rex.RexCall;
+import org.apache.calcite.rex.RexInputRef;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.sql.SqlExplainFormat;
+import org.apache.calcite.sql.SqlExplainLevel;
+import org.apache.calcite.sql.SqlKind;
+import org.apache.calcite.sql.type.SqlTypeName;
+import org.apache.commons.lang.Validate;
+import org.apache.samza.SamzaException;
+import org.apache.samza.operators.KV;
+import org.apache.samza.operators.MessageStream;
+import org.apache.samza.serializers.JsonSerdeV2;
+import org.apache.samza.serializers.KVSerde;
+import org.apache.samza.serializers.Serde;
+import org.apache.samza.sql.data.SamzaSqlCompositeKey;
+import org.apache.samza.sql.data.SamzaSqlRelMessage;
+import org.apache.samza.sql.interfaces.SourceResolver;
+import org.apache.samza.storage.kv.RocksDbTableDescriptor;
+import org.apache.samza.table.Table;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static org.apache.samza.sql.data.SamzaSqlCompositeKey.*;
+
+
+/**
+ * Translator to translate the LogicalJoin node in the relational graph to the corresponding StreamGraph
+ * implementation.
+ * Join is supported with the following caveats:
+ *   0. Only local tables are supported. Remote/composite tables are not yet supported.
+ *   1. Only stream-table joins are supported. No stream-stream joins.
+ *   2. Only Equi-joins are supported. No theta-joins.
+ *   3. Inner joins, Left and Right outer joins are supported. No cross joins, full outer joins or natural joins.
+ *   4. Join condition with a constant is not supported.
+ *   5. Compound join condition with only AND operator is supported. AND operator with a constant is not supported. No
+ *      support for OR operator or any other operator in the join condition.
+ *   6. Join condition with UDFs is not supported. Eg: udf1(a.key) = udf2(b.key) is not supported.
+ *
+ * It is assumed that the stream denoted as 'table' is already partitioned by the key(s) specified in the join
+ * condition. We do not repartition the table as bootstrap semantic is not propagated to the intermediate streams.
+ * Please refer SAMZA-1613 for more details on this. But we always repartition the stream by the key(s) specified in
+ * the join condition.
+ */
+class JoinTranslator {
+
+  private static final Logger log = LoggerFactory.getLogger(JoinTranslator.class);
+  private int joinId;
+  private SourceResolver sourceResolver;
+
+  JoinTranslator(int joinId, SourceResolver sourceResolver) {
+    this.joinId = joinId;
+    this.sourceResolver = sourceResolver;
+  }
+
+  void translate(final LogicalJoin join, final TranslatorContext context) {
+
+    // Do the validation of join query
+    validateJoinQuery(join);
+
+    boolean isTablePosOnRight = isTable(join.getRight());
+    List<Integer> streamKeyIds = new LinkedList<>();
+    List<Integer> tableKeyIds = new LinkedList<>();
```

Comment:
Question: is this under the assumption that translate() is invoked once for each task in init()? Such that the join operator here will be referring to a new instance of List objects under each task context? The upcoming serialization PR is going to change that and only invoke translate() once for each container and I wonder that this instantiation of List objects should be included in join function's init() method instead? Could you confirm the use case of these List objects? Are they supposed to be only used within the join operator in a single task context? Thanks!

Reply:
Today, SamzaSqlApplicationRunner's init calls the translate() once for each query on QueryTranslator. Within QueryTranslator translate(), we shuttle through all the joins and call translate on each of these joins. I can talk to you in person about it.